File: problems/Design Underground System/undergroundSystem.cs
public class UndergroundSystem {
    private Dictionary<string, (int, int)> _tripsStore;
    private Dictionary<int, (string, int)> _checkinStore;

    public UndergroundSystem() {
        _tripsStore = new Dictionary<string, (int, int)>();
        _checkinStore = new Dictionary<int, (string, int)>();
    }

    public void CheckIn(int id, string stationName, int t) {
        _checkinStore.Add(id, (stationName, t));
    }

    public void CheckOut(int id, string stationName, int t) {
        var (startStation, time) = _checkinStore[id];
        var key = startStation + stationName;

        if (_tripsStore.ContainsKey(key)) {
            var (sum, count) = _tripsStore[key];
            _tripsStore[key] = (sum + t - time, 1 + count);
        } else {
            _tripsStore.Add(key, (t - time, 1));
        }

        _checkinStore.Remove(id);
    }

    public double GetAverageTime(string startStation, string endStation) {
        var key = startStation + endStation;
        var (sum, count) = _tripsStore[key];

        return (double)sum / count;
    }
}

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * UndergroundSystem obj = new UndergroundSystem();
 * obj.CheckIn(id,stationName,t);
 * obj.CheckOut(id,stationName,t);
 * double param_3 = obj.GetAverageTime(startStation,endStation);
 */

File: problems/Design Underground System/undergroundSystem.java
class UndergroundSystem {

    public UndergroundSystem() {
        store = new HashMap<>();
        checkInStore = new HashMap<>();
    }

    public void checkIn(int id, String stationName, int t) {
        checkInStore.put(id, stationName + "%" + t);
    }

    public void checkOut(int id, String stationName, int t) {
        String[] items = checkInStore.get(id).split("%");
        String startStation = items[0];
        int startTime = Integer.parseInt(items[1]);
        String key = startStation + "%" + stationName;

        if (store.containsKey(key)) {
            String[] sumAndCount = store.get(key).split("%");
            int sum = Integer.parseInt(sumAndCount[0]);
            int count = Integer.parseInt(sumAndCount[1]);

            sum += t - startTime;
            ++count;

            store.put(key, "" + sum + "%" + count);
        } else {
            store.put(key, "" + (t - startTime) + "%" + 1);
        }
    }

    public double getAverageTime(String startStation, String endStation) {
        String key = startStation + "%" + endStation;
        String[] items = store.get(key).split("%");
        int sum = Integer.parseInt(items[0]);
        int count = Integer.parseInt(items[1]);

        return (double)sum / count;
    }

    private HashMap<String, String> store;
    private HashMap<Integer, String> checkInStore;
}

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * UndergroundSystem obj = new UndergroundSystem();
 * obj.checkIn(id,stationName,t);
 * obj.checkOut(id,stationName,t);
 * double param_3 = obj.getAverageTime(startStation,endStation);
 */
